content/post/codesent.md

```diff
@@ -0,0 +1,396 @@
+
+1. Debug the model locally: make sure it converges and there are no implementation faults. We debug on [e-GPUs](https://egpu.io/) at source{d}.
+2. Train the model on as much data as you have on a GPU cluster.
+3. Optionally apply a [hyperparameter optimization](https://en.wikipedia.org/wiki/Hyperparameter_optimization) algorithm to reach the best metrics in (2).
+4. Export the trained graph in "GraphDef" format (Protocol Buffers).
+5. Distribute it in an embedded or a pluggable fashion - that is, inside the application binary or separately.
+6. Apply it ("infer") with any programming language.
+
+In our case, I used the Python API to Tensorflow on a dedicated machine with 4 GPUs running two experiments
+in parallel for training and the Go API and plain CPU for inference. It is impossible to train Tensorflow
+models using the current Go bindings, but even if it was possible, I would still not do that. From my
+experience, scripting languages are superior for doing research where you rewrite code every now and then
+and need the full expressiveness and flexibility to achieve the result as soon as possible.
+At the same time Go fits the inference scenario well: we put everything into the single executable
+without dependencies and it just works. Everybody who deployed models in Python should understand
+the pain.
+
+Have you heard about [neugram](https://neugram.io/), a scripting language on top of Go runtime?
+It was created by David Crawshaw, the author of `pkg/plugin`. Neugram is still in its early ages
+and has youth related issues, but I really like it. Who knows, maybe one day we will train Tensorflow
+models using neugram - I anticipate this.
+
+The following Go code runs a Tensorflow graph:
+
+```go
+modelBytes, _ := assets.Asset("model.pb")
+graph := tf.NewGraph()
+graph.Import(modelBytes, "")
+input := graph.Operation("input").Output(0)
+output := graph.Operation("output").Output(0)
+sess, _ := tf.NewSession(graph, &tf.SessionOptions{})
+feed, _ := tf.NewTensor([][]float32{})
+result, _ := sess.Run(
+    map[tf.Output]*tf.Tensor{input: input1},
+    []tf.Output{output}, nil)
+```
+
+We embed the model using [go-bindata](https://github.com/jteeuwen/go-bindata), import the graph and
+discover input and output tensors. The new session is launched and we execute the resolved dependency
+tree to reach the output given the input. Sessions are required to allocate the hardware resources
+and manage the environment. Technically, they are bridges to the underlying Tensorflow runtime
+written in C++ as a standalone library, in case with Go this bridge is implemented with CGo.
+
+### BiDiSentiment
+
+{{% caption src="/post/codesent/gophers.png" %}}
+vmarkovtsev/BiDiSentiment logo.
+{{% /caption %}}
+
+What's left is to reveal the comment sentiment classification model. I named it "BiDiSentiment" and
+this is it's logo. BiDiSentiment is a general-purpose sentiment model, it is not specifically designed
```

Comment:
Why `BiDiSentiment`? Is there something related to Bidirectional? What does it mean?

Reply:
Oh, this is bidirectional recurrent network? use the term explicitly

Reply:
This is the name of my project. Now it's evident that I should link it to vmarkovtsev/bidisentiment :) It was in the presentation.